v7/_posts/design/2018-09-26-$expandForV4.01.md

```diff
@@ -0,0 +1,112 @@
+
+Here is an example of a simple query request and resulting response containing context url:
+
+Request url: http://services.odata.org/V4/TripPinService/People('russellwhyte')?$select=FirstName,LastName
+
+Response:
+```
+{
+    "@odata.context": "http://services.odata.org/V4/TripPinService/$metadata#People(FirstName,LastName)/$entity",
+    "@odata.id": "http://services.odata.org/V4/TripPinService/People('russellwhyte')",
+    "@odata.etag": "W/\"08D62407F53DE9ED\"",
+    "@odata.editLink": "http://services.odata.org/V4/TripPinService/People('russellwhyte')",
+    "FirstName": "Russell",
+    "LastName": "Whyte"
+}
+```
+As shown above, @odata.context annotation specifies the context url, and provides the following machine-readable description for response data:
+- It is regarding an entity from the "People" entity set,
+- It contains two property values "FirstName", "LastName".
+
+## Expanded Entity Specification for OData V4.01
+In OData V4.01, a different [format](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_ExpandedEntity) is introduced for $expand clause (for expanded entity and projected expanded entity).
+```
+10.10 Expanded Entity
+Context URL template:
+
+{context-url}#{entity-set}{/type-name}{select-list}/$entity
+
+{context-url}#{singleton}{select-list}
+
+{context-url}#{type-name}{select-list}
+
+For a 4.01 response, if a navigation property is explicitly expanded, then in addition to the non-suffixed names of any selected properties, navigation properties, functions or actions, the comma-separated list of properties MUST include the name of the expanded property, suffixed with the parenthesized comma-separated list of any properties of the expanded navigation property that are selected or expanded. If the expanded navigation property does not contain a nested $select or $expand, then the expanded property is suffixed with empty parentheses. [If the expansion is recursive for nested children, a plus sign (+) is infixed between the navigation property name and the opening parenthesis.]
+
+For a 4.0 response, the expanded navigation property suffixed with parentheses MAY be omitted from the select-list if it does not contain a nested $select or $expand, but MUST still be present, without a suffix, if it is explicitly selected.
+```
+
+## Change Summary
+According to the OData spec above, this change is for creating proper expand token in response's context url, corresponding to $expand clause in the request url. Context Url is used as response metadata and is used to control data materialization when response is received by the client.
+We need to make sure that:
+- Context url in the response generated contains correct token for the $expand clause.
+- Client can parse the context url correctly and uses the metadata for response processing.
+
+### V4.01
```

Comment:
Is this behavior consistent between 4.0 and 4.01 (desired) or only for 4.01?  Unless we think the empty parenthesis will break clients (I think we determined it did not for .NET client, other than a related bug that treated nav props wrong in general) we should just always return the parenthesized nav prop when expanded (not dependent on response version).